### ISSUE TITLE
Quick reply after open app lost message issue.

### ISSUE THREAD

**Issue by matrosovDev** [QuickBlox/quickblox-ios-sdk#309]
Hi guys, correct me if I'm wrong. As I understood when app get `- appWillResignActive` invocation we need to disconnect chat service to have ability to receive Push Notification. But the next problem here when I want to re open application and retrieve new message in the same time I can post message very quick and in the same time I am still waiting for completion receiving messages from another user. In this case I have a problem because I disconnected service and I need to reconnect with getting messages, but we will lost a message in this case because reconnecting take some time and post message operation will not trigger call back that add message to memory storage until completion:

`chatService.connectWithCompletionBlock`

**Comment by Raikerian** [comment]
Hi,

In order to not miss any messages download them from REST with filter by latest message you have in storage or latest date the chat dialog messages were updated.

**Comment by Raikerian** [comment]
Were you able to fix this issue?

**Comment by matrosovDev** [comment]
@Raikerian hey) not yet, I am not sure what did you mean about REST, I am new with QuickBlox and I just use API calls.

so as described after we used disconnectWithCompletionBlock and then use connectWithCompletionBlock it takes some time to renew session. But at the same time we can post a new message and while we wait for connectWithCompletionBlock the message will be lost. You know what I mean ) or if smth I can describe this situation with details or even make a screencast with comments.

**Comment by matrosovDev** [comment]
Seems there is only a way to create a queue of `QBChatMessage` objects and then send it when `QBChat.instance().isConnected() == true` and after `connectWithCompletionBlock`

**Comment by Raikerian** [comment]
By REST I meant QBRequest methods.
Check out our samples, especially this: https://github.com/QuickBlox/quickblox-ios-sdk/blob/master/sample-chat/sample-chat/ViewController/ChatViewController.m#L129

**Comment by matrosovDev** [comment]
It called when chat did connect.
- (void)chatServiceChatDidConnect:(QMChatService *)chatService;

 @param completion   The block which informs whether a chat did connect or not. nil if no errors.
- (void)connectWithCompletionBlock:(QBChatCompletionBlock)completion;

seems there is also a problem that app finish completion block just after we get chatServiceChatDidConnect. But as I saw in comments looks like they are for the same purposes.

**Comment by iSevenDays** [comment]
Hi matrosovDev,

chatServiceChatDidConnect is called by chatDidConnect callback

chatDidConnect happens before connectWithCompletionBlock

Scheme is below:
QB connected to chat
chatDidConnect
chatServiceChatDidConnect
completionBlock (connectWithCompletionBlock)

Before doing connect to chat you should store your messages in a queue.

Then right after connection to chat you can send them.

Thanks

**Comment by matrosovDev** [comment]
Hey Anton, thanks for answer, yea I have a queue of messages and it works for private chat, but there is an issue with group chat. in `connectWithCompletionBlock:(QBChatCompletionBlock)completion` I start sending all saved messages from a queue. it works perfect for private chat but something for a public.

**Comment by iSevenDays** [comment]
matrosovDev, please firstly check that you are joined to group chat dialog.

You can send messages to group chat dialog only after join.

http://quickblox.com/developers/SimpleSample-chat_users-ios#Chat_in_group_dialog

**Comment by Raikerian** [comment]
Were you able to fix this?

**Comment by matrosovDev** [comment]
Hey guys, I will check, I've created an array where I store my local message and then when completionBlock is ready I try to send them in a cycle. Everething works good for private chat, but for group seems I have some problem, I need to check maybe it's some local issue.

Thanks for help!!!

**Comment by Raikerian** [comment]
It could be the problem, when you are trying to send messages in group chat, but haven't joined them yet. Are you using QMServices?
If not you should perform join in a loop for all group dialogs before sending any message in it.

**Comment by matrosovDev** [comment]
Hey) I use this one  `QMServicesManager.instance().chatService.connectWithCompletionBlock {`

**Comment by Raikerian** [comment]
QMServices joining to group dialogs automatically after chat login (see QMServicesManager.m chatDidConnect delegate method). You can disable it and perform join whenever you want, so you can send messages in group chat after join was done successfully.

**Comment by matrosovDev** [comment]
Ok let me add few steps here:

1. App received - `UIApplicationWillResignActiveNotification` notification and after that application will be closed. I invoke this method below after app received this notification: `QMServicesManager.instance().chatService.disconnectWithCompletionBlock { (error) in` to disconnect all chats.

2. I reopen app and we receive new notification `UIApplicationDidBecomeActiveNotification` and here I try to connect all chats with a block `QMServicesManager.instance().chatService.connectWithCompletionBlock`

3. After `QMServicesManager.instance().chatService.connectWithCompletionBlock` was invoked I try to join all dialogs using:

`groupChat.quickBloxDialog!.joinWithCompletionBlock`

4. In completion I am staring sending messages form local array in cycle: `QMServicesManager.instance().chatService.sendMessage`

5. I checked this method `[dialog sendMessage:message completionBlock:^(NSError *error) {`and it works without error after sending a message and push back with completion, but message was lost.

**Comment by Raikerian** [comment]
What is your SDK version?
If the message was sent without any errors then you are probably haven't save it. Try load it from REST (using QBRequest methods).

**Comment by matrosovDev** [comment]
I guess the problem that the isJoined is FALSE before sending message, but I don't know how to reconnect to dialog or chat, so it does not work if I sent message immediately, but it starts work in a sec without any issues.

**Comment by Raikerian** [comment]
Please specify your Quickblox iOS SDK version (you can see it in Quickblox.h file).
If isJoined is False in completion block, print and error of it.

**Comment by matrosovDev** [comment]
frameworkVersion 2.6.3

**Comment by Raikerian** [comment]
We have performed a lot of updates for dialog joining and it is available since 2.6.5. Please consider updating your SDK. 2.6.5 and 2.6.6 contains only fixes in the core and will not affect any public interface usage.
That might be a fix for your joining problems.

**Comment by matrosovDev** [comment]
`print(publicChatRoom.quickBloxDialog?.isJoined())     // FALSE`
        
       ` publicChatRoom.quickBloxDialog?.joinWithCompletionBlock({ (error) in`
          
          `print(publicChatRoom.quickBloxDialog?.isJoined())     // FALSE`
          
        `})`

**Comment by Raikerian** [comment]
Try updating your SDK for now.

**Comment by matrosovDev** [comment]
Ok I will try, thanks, I will message to you back soon.

**Comment by matrosovDev** [comment]
This log for  frameworkVersion 2.6.5 

`publicChatRoom.quickBloxDialog?.joinWithCompletionBlock({ (error) in`
          
         ` print(publicChatRoom.quickBloxDialog?.isJoined())   // FALSE`
          
          `publicChatRoom.startSendingMessagesFromQueue();`
       ` })`

Seems the same result

**Comment by Raikerian** [comment]
Print the error variable.

**Comment by matrosovDev** [comment]
Optional(Error Domain=com.quickblox.xmpp Code=-1 "Cannot create/join room when already creating/joining/joined." UserInfo={NSLocalizedDescription=Cannot create/join room when already creating/joining/joined.}

**Comment by matrosovDev** [comment]
this is the full method with comments next to prints lines

http://codepaste.net/9aoijz

**Comment by Raikerian** [comment]
Have you removed automatic group dialog joining from QMServices? If not you should do it first. Cause you are performing dialog joining multiple times.

**Comment by matrosovDev** [comment]
Which delegate a call back can I use when dialog is connected, instead of blocks for example?

**Comment by Raikerian** [comment]
There is no delegate for such thing. If you are joining all group dialogs in loop, you can use dispatch groups to get notified that all dialogs were joined.

**Comment by matrosovDev** [comment]
yea but if you see seems error says that I already connected or join but isJoined still return false for instance.

**Comment by Raikerian** [comment]
You haven't answered my question, have you removed dialog joining code in QMServices? (QMServicesManager, remove all selector calls from chatDidConnect and chatDidReconnect delegates).

**Comment by matrosovDev** [comment]
no they are still there in the pods

**Comment by Raikerian** [comment]
Remove them first. Cause you are trying to perform your own dialog joining while there is already join in progress from QMServices.

**Comment by matrosovDev** [comment]
oh yea perfect ) should I keep those method deleted or will you update it in cocoa pods, it works for me now in that block `publicChatRoom.quickBloxDialog?.joinWithCompletionBlock({ (error)` and return TRUE for isJoined

thanks!

**Comment by Raikerian** [comment]
We will release update  for QMServices soon, where it will have user friendly property to disable those automatic joins without any code removal. For now you are good to go and can keep it as it is.

**Comment by matrosovDev** [comment]
ok cool) thank you for help!!!!

**Comment by Raikerian** [comment]
No problems, if you are still having any other troubles feel free to create separate issue :)
